java/arcs/core/analysis/AbstractValue.kt

```diff
@@ -0,0 +1,86 @@
+ * general. The interface here is designed to be general enough to allow abstract domains that are
+ * not lattices (https://en.wikipedia.org/wiki/Lattice_(order)). To that extent, we only expect
+ * the implementing class to provide an [isEquivalentTo] method that returns true if two abstract
+ * values are semantically equivalent. The [isEquivalentTo] method is also used to determine if we
+ * have reached a fixpoint when using iterative algorithms to solve the data flow equations for the
+ * problem at hand (https://en.wikipedia.org/wiki/Data-flow_analysis). For abstract domains that are
+ * lattices, the [isEquivalentTo] method can be used to compute the partial order as follows:
+ *    `a.isLessThanEqual(b)` <=> `b.isEquivalentTo(a.join(b))` [for join semi-lattice]
+ *    `a.isLessThanEqual(b)` <=> `a.isEquivalentTo(a.meet(b))` [for meet semi-lattice]
+ *
+ * For more details on the theory behind abstract interpretation, consult the following papers:
+ *   - Cousot, Patrick; Cousot, Radhia. "Abstract Interpretation: A Unified Lattice Model for
+ *     Static Analysis of Programs by Construction or Approximation of Fixpoints", In Principles
+ *     of Programming Languages (POPL), 1977.
+ *   - Cousot, Patrick; Cousot, Radhia. Abstract interpretation frameworks. Journal of Logic and
+ *     Computation, 2(4):511—547, August 1992.
+ */
+interface AbstractValue<V : AbstractValue<V>> {
+    /**
+     * Returns true if the two values are semantically equivalent.
+     *
+     * Note that this is semantic equivalence and not structural equality. This method is
+     * used to determine if we have reached a fix point. For some domains, this may simply
+     * be structural equality.
+     */
+    infix fun isEquivalentTo(other: V): Boolean
+
+    /**
+     * Returns true if the instance represents `Bottom`. `Bottom` is the lowest value in the domain
+     * lattice or represents an empty set of concrete values.
+     */
+    fun isBottom(): Boolean
+
+    /**
+     * Returns true if the instance represents `Top`. `Top` is the greatest value in the domain
+     * lattice or represents the universe of concrete values.
+     */
+    fun isTop(): Boolean
+
+    /** Returns the least upper bound of the values for lattices or a widened value. */
+    infix fun join(other: V): V
+
+    /** Returns the greatest lower bound of the values for lattices or narrowed value. */
+    infix fun meet(other: V): V
+
+    /**
+     * Returns the widened value.
+     *
+     * A widening operator is used to ensure that the increasing fixpoint computation terminates for
+     * domains with infinite ascending chains in its lattice or even to accelerate fixpoint
+     * computations (at the cost of precision) for domains with no infinite ascending chains. For
+     * lattices of finite height, this can be the same as join.
+     *
+     * Also, note that the widen operator is not commutative.
+     */
+    infix fun widen(other: V) = (this join other)
```

Comment:
how is widen different than join?

Reply:
It is needed when the underlying lattice has an infinite (or a long) ascending chain. One example for such a lattice is the interval domain, where the elements are of the form `[a, b]` and represent a set of integers `{x | a <= x <= b}`. (See [Lattice for the Interval Domain](https://www.researchgate.net/profile/Davidson_Boccardo/publication/259176840/figure/fig3/AS:654363768942598@1533023888494/The-Interval-abstract-domain.png).)  For ranges `[1,2]` and `[1,3]`, we have the following for join and widen:
- `[1,2] join [1,3] = [1,3]`
- `[1,2] widen [1,3] = [1, Int.MAX_VALUE]`

Intuitively (but informally), the widen operator allows us to extrapolate the pattern of observed changes to the state. For `[1,2] widen [1,3]`, the widen operator assumes that the upper bound would keep increasing and extrapolates the upper bound to `Int.MAX_VALUE`. Consider the following example:
 ```
 val i: Int = 0
 while (...) {
    i = i + 1 
 }
```

Reply:
Out of curiosity, does such expansion shows up when analysing recipes?

Reply:
Not for the initial use case, but if we want to reason about integer predicates in refinement types (for example), we will need it. 

Now that I think about this, we can remove the `widen` and `narrow` operations to keep things simple for now and add it later when needed.  Further, the first version of the fixpoint iterator won't be using widen and narrow anyway. We can easily do these operators back to the interface in a backward compatible way because  `widen` and `narrow` can default to `join` and `meet`,  respectively. 

Would you prefer that I remove it?

Reply:
I would personally remove it if it's not needed, though I will leave it to you.

Reply:
Removed it to keep things simple for now.